### PR TITLE
Added a directory walk to find peak annotation files

### PR DESCRIPTION
**Pull request by hepcat72**
**NOTE**: This PR is a merge into main.  When this PR is approved, the same commit will be applied to `release/1_0_2_manuscript` from this branch, into which this commit was cherry-picked: `dev/peak_annot_paths_manuscript102` (and updated after addressing review items).  (I'm still not used to main not being the bleeding edge.)

## What

- [GREATS-251](https://princeton-university.atlassian.net/browse/GREATS-251)

Use the same strategy of directory-walking (that is used for locating the mzXML files) to identify the peak annotation file paths in the PeakAnnotationFilesLoader.

Note, this will not update the study doc contents and will assume each file name is unique (which is already assumed/required by other code).

## Why

1 required step for loading (excluding changes required by errors, such as resolving mzXML file ambiguities) is the addition of the study-directory-relative paths to the Peak Annotation Files sheet.  This is a labor-intensive step when there are lots of files.

## How

Added a directory walk to find peak annotation files.

Details:

- Added exception classes: `MultipleMatchingPeakAnnotationFilesSummary` and `MultipleMatchingPeakAnnotationFiles`
- `PeakAnnotationFilesLoader`
  - Added methods:
    - `find_annot_file`
    - `map_potential_peak_annot_files`
  - Made call to `find_annot_file` from `get_file_and_format`.
  - Made call to `map_potential_peak_annot_files` from the constructor.

## Tests

- Tests added
- CI tests pass

## Security Concerns

None

## Others

There is a separate ticket that will handle this via a submission interface refactor that will allow users to drop a study folder, but until that larger effort is implemented (which solves more issues than this issue).

### CODE REVIEW

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 1]
```diff
@@ -305,9 +306,14 @@ def get_file_and_format(self, row):
...
-                    # users to use.
-                    filepath = os.path.join(study_dir, filepath_str)
+                    # Try and find the file (this will buffer an error and return None if multiple files with the same
+                    # name are found)
+                    tmp_filepath = self.find_annot_file(filepath_str, dir=study_dir)
+                    if tmp_filepath:
+                        filepath = tmp_filepath
+                    else:
+                        self.add_skip_row_index()
+                        return filename, None, format_code
```
Comment (hepcat72) [inline DataRepo/loaders/peak_annotation_files_loader.py:319]: C, At this point in the conditionals, we know that the supplied peak annotation file does not have accurate path information, so this code change adds a call to the new `find_annot_file` method to try and find the file.  If it is not found, the row in the `Peak Annotation Files` sheet is marked to be skipped and the return from this method (`get_file_and_format`) indicates that the file/format could not be found/determined.

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 2]
```diff
@@ -474,9 +481,8 @@ def load_peak_annotations(
             or format_code not in PeakAnnotationsLoader.get_supported_formats()
         ):
             self.buffer_infile_exception(
-                (
-                    f"Skipping load of peak annotations file '{filename}'.  Unrecognized format code: {format_code}.  "
-                    f"Must be one of {PeakAnnotationsLoader.get_supported_formats()}."
+                InvalidPeakAnnotationFileFormat(
+                    format_code, PeakAnnotationsLoader.get_supported_formats(), filepath
```
Comment (hepcat72) [inline DataRepo/loaders/peak_annotation_files_loader.py:487]: C. I just happened to notice that this was wrong.  The first argument to `buffer_infile_exception` is supposed to be an exception object, but it was passing a `str`.  I believe this was never encountered because this is always fixed by the user during validation.  This loader is not run during validation - only during loading, at which point the user had to have fixed it.  The validation page never gets here because it doesn't run this loader.  It takes too long.

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 3]
```diff
@@ -675,3 +681,64 @@ def get_study_dir(self):
...
+        Args:
+            supplied_files_with_opt_path (str): A peak annotation file with optional path (derived from the Peak
+                Annotation Files sheet).
+            dir (Optional[str]) [current directory]: A directory under which mzXML files reside (in subdirectories).
+        Exceptions:
+            None
+        Returns:
+            annot_filepath (Optional[str]): A peak annotation filepath if one was found, the supplied_file_with_opt_path
+                if not found, or None if multiple matching files were found (i.e. error).
+        """
```
Comment (ro9002) [inline DataRepo/loaders/peak_annotation_files_loader.py:702]: 'find_annot_file' has multiple return modes with different semantics, this is not type-safe. Why is "None" used to model an error? Rob, please use an explicit modeling instead since there are 3 clear outcomes. For instance: 
```
class AnnotFileResult:
    status: Literal["FOUND", "NOT_FOUND", "MULTIPLE"]
    path: Optional[str]
```

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 4]
```diff
@@ -675,3 +681,64 @@ def get_study_dir(self):
...
+        supplied_dirpath, supplied_filename = os.path.split(supplied_file_with_opt_path)
+
+        # Walk the directory
+        for dir_path, _, fileset in os.walk(dir):
+            for filename in fileset:
+                # If the current file matches one of the supplied files
+                if filename == supplied_filename:
+                    filepath = os.path.join(dir_path, filename)
+                    annot_filepath = filepath
+                    matching_annot_files.append(filepath)
```
Comment (ro9002) [inline DataRepo/loaders/peak_annotation_files_loader.py]: The loop continues even after len(matching_annot_files) > 1, causing unnecessary traversal. Break the os.walk once multiple matches are found to improve efficiency and scalability.

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 5]
```diff
@@ -675,3 +681,64 @@ def get_study_dir(self):
...
             return os.getcwd()
         return os.path.dirname(os.path.abspath(self.file))
+
+    def find_annot_file(self, supplied_file_with_opt_path: str, dir=None):
+        """Given a supplied file name with optional path information, return it with its filepath.
+
+        Args:
+            supplied_files_with_opt_path (str): A peak annotation file with optional path (derived from the Peak
+                Annotation Files sheet).
+            dir (Optional[str]) [current directory]: A directory under which mzXML files reside (in subdirectories).
```
Comment (ro9002) [inline DataRepo/loaders/peak_annotation_files_loader.py]: There is a python built-in called "dir()". 'dir' should be renamed to avoid confusion.

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 6]
```diff
@@ -675,3 +681,64 @@ def get_study_dir(self):
...
+        if len(matching_annot_files) > 1:
+            annot_filepath = None
+            self.buffer_infile_exception(
+                MultipleMatchingPeakAnnotationFiles(matching_annot_files)
+            )
+        elif annot_filepath is None:
+            # Return the supplied file name if not found at all (the caller will deal with the ensuing error)
+            annot_filepath = supplied_file_with_opt_path
+        elif annot_filepath and supplied_dirpath not in ["", "."]:
+            # Buffer a warning that the supplied path was wrong, (but the file was found).
```
Comment (ro9002) [inline DataRepo/loaders/peak_annotation_files_loader.py:733]: Please explain the rationale here?

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 7]
```diff
@@ -675,3 +681,64 @@ def get_study_dir(self):
...
+                    f"The peak annotation filepath supplied '{supplied_file_with_opt_path}' was incorrect but a file "
+                    "matching this filename was found in the study directory, so this is just a warning.  Correct the "
+                    "Peak Annotation Files sheet to eliminate this warning."
+                ),
+                is_error=False,
+                is_fatal=self.validate,
+                column=self.headers.FILE,
+            )
+
+        return annot_filepath
```
Comment (ro9002) [inline DataRepo/loaders/peak_annotation_files_loader.py]: The current logic uses both annot_filepath and matching_annot_files to represent the same underlying state. Specifically:
When len(matching_annot_files) > 1, the function sets annot_filepath = None
When len(matching_annot_files) == 1, annot_filepath is set to matching_annot_files[0]
This introduces redundant tracking and makes the control flow harder to follow. Please simplify the implementation by using a single source of truth (i.e., matching_annot_files) and deriving the return value directly from it. This will improve readability, clarity, and maintainability.

File: DataRepo/utils/exceptions.py [hunk 8]
```diff
@@ -5842,6 +5843,64 @@ def __init__(self, filename):
...
+    TraceBase requires that peak annotation filenames be globally unique to avoid ambiguities when sharing or
+    referencing data files.
+    """
+
+    SummarizerExceptionClass = MultipleMatchingPeakAnnotationFilesSummary
+
+    def __init__(self, annot_files: List[str], **kwargs):
+        nltt = "\n\t\t"
+        message = (
+            f"A peak annotation filename from %s had multiple matching files found in the study directory:\n"
```
Comment (ro9002) [inline DataRepo/utils/exceptions.py:5895]: What is the format for "%s" ?

File: DataRepo/utils/exceptions.py [hunk 9]
```diff
@@ -5842,6 +5843,64 @@ def __init__(self, filename):
...
+    To resolve this issue, either delete the duplicates (recommended) or specify the path (relative to the study
+    directory) to the desired file in the Peak Annotation Files sheet.
+
+    TraceBase requires that peak annotation filenames be globally unique to avoid ambiguities when sharing or
+    referencing data files.
+    """
+
+    def __init__(self, exceptions: List[MultipleMatchingPeakAnnotationFiles]):
+        # Collect all the files in a 2D dict
+        files_dict: Dict[str, Dict[str, int]] = defaultdict(lambda: defaultdict(int))
```
Comment (ro9002) [inline DataRepo/utils/exceptions.py]: What's the rationale for "Dict[str, Dict[str, int]]" ?

File: DataRepo/loaders/peak_annotation_files_loader.py [hunk 10]
```diff
@@ -675,3 +681,64 @@ def get_study_dir(self):
...
+                    annot_filepath = filepath
+                    matching_annot_files.append(filepath)
+
+        # If there are multiple files with this name (which is not allowed)
+        if len(matching_annot_files) > 1:
+            annot_filepath = None
+            self.buffer_infile_exception(
+                MultipleMatchingPeakAnnotationFiles(matching_annot_files)
+            )
+        elif annot_filepath is None:
```
Comment (ro9002) [inline DataRepo/loaders/peak_annotation_files_loader.py]: Why is this error buffered?

**Review by ro9002 — CHANGES_REQUESTED** [review]
Thank you, Rob, for your contribution. There are several important issues that need to be addressed before this PR can be approved:
- find_annot_file behavior and design: The function currently has ambiguous return semantics (mixing valid paths, fallback input values, and None) and combines file resolution with error handling. This makes the behavior harder to reason about and increases risk for callers. In addition, the search logic should be improved for clarity and efficiency.
- Code quality and maintainability: Please ensure the implementation aligns with our standards for clear function contracts, predictable behavior, variable naming, and reviewable, performant code.
- Test coverage: Additional test cases are needed to improve robustness. For example, consider edge cases such as mixed-case filenames and symbolic links, along with other filesystem variations where applicable.
- Process adherence: This change should also follow the established workflow (dev -> release -> main), including appropriate testing in a staging or production-like environment before merging (by the dev team and possibly users as well).

File: DataRepo/loaders/peak_annotation_files_loader.py (hunk 5 above)
Comment (hepcat72) [reply DataRepo/loaders/peak_annotation_files_loader.py]: Huh.  Oh right.  I use that a lot in the django shell when debugging.  Funny that didn't occur to me when developing.  I think I've used `dir` as a variable elsewhere too.  Definitely should not use that reserved word as a variable name.  Thanks.

- [x] Rename `dir` argument

File: DataRepo/loaders/peak_annotation_files_loader.py (hunk 3 above)
Comment (hepcat72) [reply DataRepo/loaders/peak_annotation_files_loader.py:702]: That's an interesting suggestion.  Let me answer your question as to why I chose to use `None` (and why I return the supplied file name when not found - creating 3 return states), because the strategy here was intentional.  Normally, I would only return 1 type of value, so you're right to point this out.  I would make the same comment if I was reviewing this as someone else's code.  In fact, one of the last reviews I made on John's code, I made a similar comment.

The reason I did it this way is because I consider this fix to be a temporary fix.  I alluded to this in the **Others** section in the PR description.  This loader and the associated data sheet weren't designed to take peak annotation file names and find those files.  There already exists mechanisms in the surrounding code that handles missing files (but requires a relative path to that file).  The better fix, that I think should be implemented in the next major version (if there is one) is described in GH issue #1267.  That issue focusses on mzXML files, but applies to peak annotation files as well.  And I essentially did a spike/proof of concept for that feature over a year ago in #1442.

That solution is consistent with the overall design.

As such, I was trying to implement a fix that required the least change to the surrounding code, so that when this method is removed and 1267 is implemented, no other changes need to be made.

This method adds 2 new error types (well, one error, one warning) to handle the new cases this method creates: multiple files with the same name (which is otherwise supported in the original design) and the wrong path provided (but the file was found - making it a warning).

I was intending the returned value to allow the surrounding code to handle `FileFromInputNotFound` errors the same as before, but if I returned the supplied filename in the multiple files found case, then that error would be redundant to the new multiple files found error.

I like your suggestion though, even though I intend this fix to be temporary.  I'll see about implementing it or something like it.

- [x] Create a single return type for `find_annot_file`

File: DataRepo/loaders/peak_annotation_files_loader.py (hunk 4 above)
Comment (hepcat72) [reply DataRepo/loaders/peak_annotation_files_loader.py]: I think this would cause a potential problem for users.  If they have >2 files with the same name, and our error shows them only 2 of them, but they are the wrong 2 (because the third one is the correct choice, they would end up with incorrect data.

It's funny, because I had thought that the efficiency issue you would raise would be the multiple walk traversals (because this is called in a loop from the calling method).  I had half a mind to implement the full traversal once from the constructor and save it in an instance attribute before I started coding this fix, but since I considered this temporary code, to be removed later, and I had other priorities impending, I went with the quicker fix and the smaller code profile for later easy removal.

I will change it to:

- [x] Do the walk/traversal once and save the result for access later

File: DataRepo/loaders/peak_annotation_files_loader.py (hunk 10 above)
Comment (hepcat72) [reply DataRepo/loaders/peak_annotation_files_loader.py]: All errors are buffered, as is the overall strategy of all the loader classes (that are re-used in the validation interface).  Before we started aggregating errors, the load process took literally weeks because any time an error was raised, the curator had to fix the study doc and start over.  Running a load can take hours, depending on the side of the study.  To have to start over at each error created a ton of churn.  By switching to an aggregated errors strategy, curators could get nearly all the errors that need to be addressed in one shot, which dramatically improved the submission validating and data debugging process, and cut down on overall load time.

File: DataRepo/loaders/peak_annotation_files_loader.py (hunk 6 above)
Comment (hepcat72) [reply DataRepo/loaders/peak_annotation_files_loader.py:733]: I assume you're asking what this is checking and why I am issuing a warning instead of an error.

First, let me explain this conditional, because it's a bit indirect:

`elif annot_filepath and supplied_dirpath not in ["", "."]:`

Checking `annot_filepath` having a value indicates that a file was found.  (Note: We know that the supplied file path is wrong because this method was called.)  Checking that `supplied_dirpath` is not empty (or was normalized to the current directory) tells us that a path had been supplied.  Thus that means that the supplied path was wrong.

Since a unique file matching the supplied file was found, the problem has been averted, thus we don't issue an error.  However, we're assuming we found the correct file.  We could be making a decision here that is wrong.  Maybe the user previously had 2 files with the same name and deleted the wrong one.  Thus we issue a warning to tell the user:

```
                    f"The peak annotation filepath supplied '{supplied_file_with_opt_path}' was incorrect but a file "
                    "matching this filename was found in the study directory, so this is just a warning.  Correct the "
                    "Peak Annotation Files sheet to eliminate this warning."
```

We assume the best because a guiding principle in the development of TraceBase was to lower the barrier to submission that other metabolomics sites have set so high that users struggle to submit data.  Thus TraceBase tries to smooth the rough corners wherever possible.

File: DataRepo/loaders/peak_annotation_files_loader.py (hunk 7 above)
Comment (hepcat72) [reply DataRepo/loaders/peak_annotation_files_loader.py]: Good catch.  This happened because originally `matching_annot_files` was named `dupe_annot_files` and I was populating it only when I found multiple (when `annot_filepath` was already set).  And I'd been using a boolean as well.  I had simplified it to remove the boolean after my first pass and changed the behavior and name of the dupe variable.  I was anxious to get to the loading doc, so I apparently stopped short of catching that redundancy myself.  You've got a good eye for catching things like this.  I appreciate it.

- [x] Address `annot_filepath` redundancy

File: DataRepo/utils/exceptions.py (hunk 9 above)
Comment (hepcat72) [reply DataRepo/utils/exceptions.py]: Now that you ask, I immediately see that I could have used `Dict[str, List[str]]`.  I'd only wanted the uniqueness applied to get the secondary keys, but since these are coming from the file system, they're already inherently unique.  Good catch.  You did it again!

- [x] Change `Dict[str, Dict[str, int]]` to `Dict[str, List[str]]` (and sort while I'm at it)

File: DataRepo/utils/exceptions.py (hunk 8 above)
Comment (hepcat72) [reply DataRepo/utils/exceptions.py:5895]: It is the file location information where the peak annotation file path in question was supplied.  In this case, it is a specific row of the `Peak Annotation File` column in the `Peak Annotation Files` sheet of the supplied study doc.

Incidentally, I have an old GH issue (#914) to improve the `InfileError` superclass that fills that placeholder in.  I wrote that class before I'd learned that you can supply a dict to fill in `{DICTKEY}` in strings using `.format()` to fill in values.  I was using `%s` because I'd learned that by example from John and Lance's code, and it can be pretty well simplified if I use that strategy.  When I created that issue, I'd recently learned that that `%s` strategy is outmoded.

Comment (hepcat72) [reply DataRepo/utils/exceptions.py:5895]: Oh, you asked about the format...  The format is dynamically generated by `generate_file_location_string()` based on what info was supplied - sometimes it indicates a whole row, sometimes a whole column, sometimes a whole sheet, or even just the file.

**Comment by hepcat72** [comment]
> Thank you, Rob, for your contribution. There are several important issues that need to be addressed before this PR can be approved:
> 
> * find_annot_file behavior and design: The function currently has ambiguous return semantics (mixing valid paths, fallback input values, and None) and combines file resolution with error handling. This makes the behavior harder to reason about and increases risk for callers. In addition, the search logic should be improved for clarity and efficiency.

See responses/changes in-line.

> * Code quality and maintainability: Please ensure the implementation aligns with our standards for clear function contracts, predictable behavior, variable naming, and reviewable, performant code.

It would help me a lot if you could point me to the specific definitions of our standards that you reference, so that I can align with your expectations, specifically regarding these referenced standards:

- function contracts
- predictable behavior
- reviewable [code]
- performant code

I prefer not to assume what each of these means or even look up specific definitions which could diverge from your understanding.  I would also like to have well defined exceptions to these expectations in different contexts so that I know what to expect.

> * Test coverage: Additional test cases are needed to improve robustness. For example, consider edge cases such as mixed-case filenames and symbolic links, along with other filesystem variations where applicable.

I feel this is an out of scope case.  Supporting symlinks and filesystem edge cases are not in the acceptance criteria.  It would greatly help me to establish predicability in code reviews so that everything is known up front when a ticket is added to the board.  If new acceptance criteria arise, it should be a different ticket.  Otherwise, this represents uncontrolled change.

Aside from that, I don't feel that support for symlinks and file system variations is a likely edge case.  Adding that support piecemeal throughout the code-base would create a heterogenous and unpredictable environment.  Plus, this code, as indicated in the ticket and the PR description, is temporary, thereby reducing the value of making such a change.  See GREATS-251's reference to GREATS-291.

> * Process adherence: This change should also follow the established workflow (dev -> release -> main), including appropriate testing in a staging or production-like environment before merging (by the dev team and possibly users as well).

As stated at the top of the PR, this work will be applied to `release/1_0_2_manuscript`.  My intention was to follow the release branch strategy.  It just hasn't been a habit I've been able to make second nature yet.  Creating a separate PR would be an academic exercise, as the CI tests run upon push.

Regarding testing in a production-like environment, I feel like that is unnecessary.  This is a simple fix and there is no large scale data yet available to test it on that is not already loaded.  It is my opinion that the added CI tests are entirely sufficient and thorough for this change.  Besides, it will get tested in a very convenient case when you run it in the training.

**Review by ro9002 — CHANGES_REQUESTED** [review]
Rob,
This PR is not approved. Approval is pending the required changes. I refer you to the detailed messages/discussions in Files Changed tab.

Regarding the test coverage for the filesystem behavior. If you believe a specific case should not be supported, state that explicitly in the PR, explain the technical reason, and identify why the risk is low. No item is deferred unless that is explicitly agreed in review. The examples I provided in review, including mixed-case filenames and symlink-related behavior where applicable, were intended to illustrate the level of edge-case analysis expected when file-resolution logic is introduced. If you believe a scenario is not applicable, again address that in the PR with a concise technical rationale rather than leaving the risk itself unaddressed.

Follow the team PR SOP and release workflow. Use the required branch path (dev -> release -> main) and update this PR to match that workflow. You noted that your intent was to follow that strategy. That is the clear expected workflow under the team’s established PR process and should be followed consistently. CI passing on push does not replace the required PR and branch process.

Regarding the verification, let me clarify: testing in a production-like environment is part of release-candidate verification for the overall release once development work is complete (so once the V1.0.2 release is ready). It is not a special step for this PR alone. CI coverage is necessary, but it is not the only verification step in the team’s delivery process.

These expectations are consistent with the team’s documented PR process, written reviews in previous PRs, technical discussions in 1:1s and team's core principles. Please address the review items and post an updated revision by end of this week (April 17). If there is still confusion on any specific item, schedule time with me and we will resolve it directly.

**Comment by hepcat72** [comment]
> Rob, This PR is not approved. Approval is pending the required changes. I refer you to the detailed messages/discussions in Files Changed tab.

To document our discussion yesterday, I forgot to hit submit so you could see all my comments. I just hit submit, so you should be able to see that I have responded to every issue, however you should have also been able to see that I implemented changes addressing your issues.

I would also like to make a terminology clarification regarding your use of "required changes". What is required, as is indicated in the SOP is that we agree on a resolution to every review issue. Sometimes that is a "change", sometimes it is a ticket, and sometimes is is an agreement that an issue is invalid. I would just like to be precise with our terminology, if you don't mind.

> Regarding the test coverage for the filesystem behavior. If you believe a specific case should not be supported, state that explicitly in the PR, explain the technical reason, and identify why the risk is low.

You should be able to see my response that does this now, though I will point out that there was no listed requirement in the ticket to support links or miscellaneous file systems. The only exception to explicitly stating a requirement in a ticket would be if there is a general requirement in the project documentation. Supporting links and other file systems has never been a general requirement in this project. Without that, a ticket must clearly state such a requirement. If we follow this process, work will go more smoothly and predictably.

> No item is deferred unless that is explicitly agreed in review.

I don't understand why you are stating this. Do you have any reason to believe that I believe I have the power to unilaterally defer a review item? We've talked about this many times and you have reviewed the independent code review guidelines that I shared with you on this very topic that clearly states that the process requires that we come to an agreement and I believe that review comments block merge. I have never intentionally merged work that had unresolved blocking review issues. And as I reiterated yesterday, the reviewer owns the merge and the author owns the code. I have been consistent in this view, as evidenced by multiple discussions. Governing this process was my explicit job as Configuration manager and configuration control board leader in my team at Los Alamos. [Anticipating your response, I only cite this to point out that this is a long held view and not that I expect we follow the specific process we had at Los Alamos,] You complained yesterday about how many times you feel you have to remind me about process expectations but I feel the same way. I don't understand why I must reiterate my understanding and firm belief in this review process. So this frustration goes both ways.

> The examples I provided in review, including mixed-case filenames and symlink-related behavior where applicable, were intended to illustrate the level of edge-case analysis expected when file-resolution logic is introduced. If you believe a scenario is not applicable, again address that in the PR with a concise technical rationale rather than leaving the risk itself unaddressed.

I did address it and you should be able to see that since I now have hit submit on my comments. This has happened before, if you recall, where I forgot to hit submit on a series of comments and at that time, I told you that it is always my intention to not only address every issue but also reply to every comment, even if it is just "done" and the issue is just "typo". And at that time, I requested that if you ever saw that I rerequested review and hadn't appeared to reply to any comments, to let me know. So, to remind you again, if that happens, please assume the intent on my part was to reply to every comment and that if I appear to have not done that, then it is a simple oversight on my part to hit that submit button. By the way, to provide some context and clear up potential confusion, the reason I treat my comments as a review instead of individual replies is that sometimes replying to later comments prompts me to edit my reply to previous comments. I don't know if when a series of review comments are pending, whether my clicking any resolve button is visible by you. I only say this to clarify that if it ever does appear I have hit the resolve button without having made a comment, then it is extremely likely I forgot to hit submit on the comment series.

> Follow the team PR SOP and release workflow. Use the required branch path (dev -> release -> main) and update this PR to match that workflow. You noted that your intent was to follow that strategy. That is the clear expected workflow under the team’s established PR process and should be followed consistently. CI passing on push does not replace the required PR and branch process.

I had addressed this via reply which you should now be able to see.

> Regarding the verification, let me clarify: testing in a production-like environment is part of release-candidate verification for the overall release once development work is complete (so once the V1.0.2 release is ready). It is not a special step for this PR alone.

Thanks for that clarification. That is helpful.

> CI coverage is necessary, but it is not the only verification step in the team’s delivery process.
> 
> These expectations are consistent with the team’s documented PR process, written reviews in previous PRs, technical discussions in 1:1s and team's core principles. Please address the review items and post an updated revision by end of this week (April 17). If there is still confusion on any specific item, schedule time with me and we will resolve it directly.

I will request rereview RN since I hit submit on my comments and all issues have been addressed. (And just as a reminder to be clear, by "addressed", as explained in my independent review guidelines document that you have read but we've not adopted, I mean that I either made a change, made a ticket, or replied explaining why the issue should not be addressed either due to scope or validity concerns.

**Review by ro9002 — CHANGES_REQUESTED** [review]
@hepcat72  I reviewed PR #1740. There have been no new commits since my initial review. So the required updates have not been completed. This PR is not approved.
Post an updated revision with the required commits by end of day today. Any item you believe does not require a code change must be documented in the PR with a concise technical rationale and risk assessment. Do not re-request review until the updates are posted.

File: DataRepo/utils/exceptions.py (hunk 9 above)
Comment (ro9002) [reply DataRepo/utils/exceptions.py]: What about Set[str] ?

**Comment by hepcat72** [comment]
Closed because you cannot change (as had been suggested) the target branch to the release branch because the release branch is 332 commits behind main and this branch was branched off main, so merging this branch into `release/1_0_2_manuscript` would add the work from those 332 commits *that we do not want in the release branch* into the release branch.  I.e. it's not a matter of simply resolving conflicts.  It's a matter of the fact that this branch differs from the release and merging merges all those differences.
